third_party/terraform/utils/k8s_operation.go

```diff
@@ -0,0 +1,113 @@
+package google
```

Comment:
I don't think this is a generic representation of asynchronous resources in K8S, I think it's a representation of the rules of an asynchronous knative resource. [knative Error Signalling](https://github.com/knative/docs/blob/master/docs/serving/spec/knative-api-specification-1.0.md#error-signalling) describes the assumptions made here but [K8S Spec and Status](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#object-spec-and-status) and [K8S Typical Status Properties](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) don't.

Reply:
Ah great. I think I may have over generalized the knative design to the k8s. For context I was basing my assumption on the fact that both the [pod conditions](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) and the [node conditions](https://kubernetes.io/docs/concepts/architecture/nodes/#condition) make explicit reference to the ready condition. 
Additionally discussions in [some kubernetes design issues](https://github.com/kubernetes/kubernetes/issues/7856) seem to indicate that the status.conditions block will continue to be a thing and that the ready condition is the most important condition (personal interpretation of the discussion).

Also as I mentioned in the PR overview - the assumption on ready:true is part of the minimal implementation until there are more resources that require this style and we can generalize. In the future if we think that the terminal status of a resource needs to be represented on a per resource basis we can add the functionality to the async object so that it can pass in this information. This could be thrown into a post_update/create but by using those we don't get the benefit of any of the shared functionality of async such as timeout overloading.

Regarding your question - are you saying that this is better off as a knative operation rather than a generalized kubernetes?

Reply:
My read of the linked GH thread was the `Ready` shouldn't even be a `Condition`: https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-335687733

I developed my thoughts more in https://github.com/GoogleCloudPlatform/magic-modules/pull/2716#issuecomment-555675554.

I think that we can create a stronger abstraction by treating resource-field-async and LRO-async as different things. While Cloud Run is resource-field-async, and typical resources are LRO-async, GKE supports both and I think there are a couple other resources in the same place.

I think the `Async` class is badly named, and should be named `AsyncOperation` or similar. We can add `AsyncStatus` as a separate block in resource definitions, which supplies values a provider can use to determine the state of a resource.

I think that with `AsyncStatus` GCE instance states could be represented, too. They've got LROs and a state field that indicates whether the instance is running or not.